### ISSUE TITLE
Panic in src/lsm/manifest.zig:317

### ISSUE THREAD

**Issue by eatonphil** [tigerbeetle/tigerbeetle#175]
Running [this script](https://gist.github.com/eatonphil/3bceb01e603ea0edcae16217c592c2fb) against the latest main commit I got:

```
./tigerbeetle start --addresses=3000 0_0.tigerbeetle
info(io): opening "0_0.tigerbeetle"...
info(main): 0: cluster=0: listening on 127.0.0.1:3000
info(message_bus): connection from client 206272961074970794762714954581878172245
thread 2523663 panic: reached unreachable code
/home/phil/tb/tigerbeetle/zig/lib/std/debug.zig:225:14: 0x26920b in std.debug.assert (tigerbeetle)
    if (!ok) unreachable; // assertion failure
             ^
/home/phil/tb/tigerbeetle/src/lsm/manifest.zig:317:23: 0x42a7ee in lsm.manifest.ManifestType(lsm.table.TableType(u128,lsm.groove.IdTreeValue,lsm.groove.IdTreeValue.compare_keys,lsm.groove.IdTreeValue.key_from_value,554112867134706473364364839029663282043,lsm.groove.IdTreeValue.tombstone,lsm.groove.IdTreeValue.tombstone_from_key),storage.Storage).assert_no_invisible_tables (tigerbeetle)
                assert(it.next() == null);
                      ^
/home/phil/tb/tigerbeetle/src/lsm/tree.zig:873:57: 0x404f46 in lsm.tree.TreeType(lsm.table.TableType(u128,lsm.groove.IdTreeValue,lsm.groove.IdTreeValue.compare_keys,lsm.groove.IdTreeValue.key_from_value,554112867134706473364364839029663282043,lsm.groove.IdTreeValue.tombstone,lsm.groove.IdTreeValue.tombstone_from_key),storage.Storage,[]const u8{84,114,97,110,115,102,101,114,46,105,100}).checkpoint (tigerbeetle)
                tree.manifest.assert_no_invisible_tables(snapshot);
                                                        ^
/home/phil/tb/tigerbeetle/src/lsm/groove.zig:890:34: 0x3e99cb in lsm.groove.GrooveType(storage.Storage,tigerbeetle.Transfer,(struct state_machine.struct:40:14 constant)).checkpoint (tigerbeetle)
            groove.ids.checkpoint(Join.tree_callback(.ids));
                                 ^
/home/phil/tb/tigerbeetle/src/lsm/forest.zig:200:62: 0x3c202c in lsm.forest.ForestType(storage.Storage,(struct state_machine.struct:47:49 constant)).checkpoint (tigerbeetle)
                @field(forest.grooves, field.name).checkpoint(Join.groove_callback(field.name));
                                                             ^
/home/phil/tb/tigerbeetle/src/state_machine.zig:382:35: 0x3a3117 in state_machine.StateMachineType(storage.Storage).checkpoint (tigerbeetle)
            self.forest.checkpoint(checkpoint_finish);
                                  ^
/home/phil/tb/tigerbeetle/src/vsr/replica.zig:2531:46: 0x386e72 in vsr.replica.ReplicaType(state_machine.StateMachineType(storage.Storage),message_bus.MessageBusType(vsr.ProcessType.replica),storage.Storage,time.Time).commit_op_compact_callback (tigerbeetle)
                self.state_machine.checkpoint(commit_op_checkpoint_state_machine_callback);
                                             ^
/home/phil/tb/tigerbeetle/src/state_machine.zig:374:21: 0x3a305a in state_machine.StateMachineType(storage.Storage).compact_finish (tigerbeetle)
            callback(self);
                    ^
/home/phil/tb/tigerbeetle/src/lsm/forest.zig:169:37: 0x3e94df in lsm.forest.ForestType(storage.Storage,(struct state_machine.struct:47:49 constant)).JoinType(lsm.forest.JoinOp.compacting).groove_callback().groove_cb (tigerbeetle)
                            callback(forest);
                                    ^
/home/phil/tb/tigerbeetle/src/lsm/groove.zig:848:37: 0x40350d in lsm.groove.GrooveType(storage.Storage,tigerbeetle.Transfer,(struct state_machine.struct:40:14 constant)).JoinType(lsm.groove.JoinOp.compacting).tree_callback().tree_cb (tigerbeetle)
                            callback(groove);
                                    ^
/home/phil/tb/tigerbeetle/src/lsm/tree.zig:848:21: 0x464318 in lsm.tree.TreeType(lsm.table.TableType(lsm.composite_key.CompositeKey(u64),lsm.composite_key.Value,lsm.composite_key.CompositeKey(u64).compare_keys,lsm.composite_key.CompositeKey(u64).key_from_value,(struct lsm.composite_key.CompositeKey(u64) constant),lsm.composite_key.CompositeKey(u64).tombstone,lsm.composite_key.CompositeKey(u64).tombstone_from_key),storage.Storage,[]const u8{84,114,97,110,115,102,101,114,46,97,109,111,117,110,116}).compact_manifest_callback (tigerbeetle)
            callback(tree);
                    ^
/home/phil/tb/tigerbeetle/src/lsm/manifest.zig:539:21: 0x47eab3 in lsm.manifest.ManifestType(lsm.table.TableType(lsm.composite_key.CompositeKey(u64),lsm.composite_key.Value,lsm.composite_key.CompositeKey(u64).compare_keys,lsm.composite_key.CompositeKey(u64).key_from_value,(struct lsm.composite_key.CompositeKey(u64) constant),lsm.composite_key.CompositeKey(u64).tombstone,lsm.composite_key.CompositeKey(u64).tombstone_from_key),storage.Storage).manifest_log_compact_callback (tigerbeetle)
            callback(manifest);
                    ^
/home/phil/tb/tigerbeetle/src/lsm/manifest_log.zig:451:25: 0x4c233c in lsm.manifest_log.ManifestLogType(storage.Storage,lsm.manifest.TableInfoType(lsm.table.TableType(lsm.composite_key.CompositeKey(u64),lsm.composite_key.Value,lsm.composite_key.CompositeKey(u64).compare_keys,lsm.composite_key.CompositeKey(u64).key_from_value,(struct lsm.composite_key.CompositeKey(u64) constant),lsm.composite_key.CompositeKey(u64).tombstone,lsm.composite_key.CompositeKey(u64).tombstone_from_key))).compact_flush_callback (tigerbeetle)
                callback(manifest_log);
                        ^
/home/phil/tb/tigerbeetle/src/lsm/manifest_log.zig:358:25: 0x46b9a4 in lsm.manifest_log.ManifestLogType(storage.Storage,lsm.manifest.TableInfoType(lsm.table.TableType(lsm.composite_key.CompositeKey(u64),lsm.composite_key.Value,lsm.composite_key.CompositeKey(u64).compare_keys,lsm.composite_key.CompositeKey(u64).key_from_value,(struct lsm.composite_key.CompositeKey(u64) constant),lsm.composite_key.CompositeKey(u64).tombstone,lsm.composite_key.CompositeKey(u64).tombstone_from_key))).write_block (tigerbeetle)
                callback(manifest_log);
                        ^
/home/phil/tb/tigerbeetle/src/lsm/manifest_log.zig:338:37: 0x4c210e in lsm.manifest_log.ManifestLogType(storage.Storage,lsm.manifest.TableInfoType(lsm.table.TableType(lsm.composite_key.CompositeKey(u64),lsm.composite_key.Value,lsm.composite_key.CompositeKey(u64).compare_keys,lsm.composite_key.CompositeKey(u64).key_from_value,(struct lsm.composite_key.CompositeKey(u64) constant),lsm.composite_key.CompositeKey(u64).tombstone,lsm.composite_key.CompositeKey(u64).tombstone_from_key))).flush (tigerbeetle)
            manifest_log.write_block();
                                    ^
/home/phil/tb/tigerbeetle/src/lsm/manifest_log.zig:421:31: 0x47ea26 in lsm.manifest_log.ManifestLogType(storage.Storage,lsm.manifest.TableInfoType(lsm.table.TableType(lsm.composite_key.CompositeKey(u64),lsm.composite_key.Value,lsm.composite_key.CompositeKey(u64).compare_keys,lsm.composite_key.CompositeKey(u64).key_from_value,(struct lsm.composite_key.CompositeKey(u64) constant),lsm.composite_key.CompositeKey(u64).tombstone,lsm.composite_key.CompositeKey(u64).tombstone_from_key))).compact (tigerbeetle)
            manifest_log.flush(compact_flush_callback);
                              ^
/home/phil/tb/tigerbeetle/src/lsm/manifest.zig:530:42: 0x45e901 in lsm.manifest.ManifestType(lsm.table.TableType(lsm.composite_key.CompositeKey(u64),lsm.composite_key.Value,lsm.composite_key.CompositeKey(u64).compare_keys,lsm.composite_key.CompositeKey(u64).key_from_value,(struct lsm.composite_key.CompositeKey(u64) constant),lsm.composite_key.CompositeKey(u64).tombstone,lsm.composite_key.CompositeKey(u64).tombstone_from_key),storage.Storage).compact (tigerbeetle)
            manifest.manifest_log.compact(manifest_log_compact_callback);
                                         ^
/home/phil/tb/tigerbeetle/src/lsm/tree.zig:819:34: 0x447c29 in lsm.tree.TreeType(lsm.table.TableType(lsm.composite_key.CompositeKey(u64),lsm.composite_key.Value,lsm.composite_key.CompositeKey(u64).compare_keys,lsm.composite_key.CompositeKey(u64).key_from_value,(struct lsm.composite_key.CompositeKey(u64) constant),lsm.composite_key.CompositeKey(u64).tombstone,lsm.composite_key.CompositeKey(u64).tombstone_from_key),storage.Storage,[]const u8{84,114,97,110,115,102,101,114,46,97,109,111,117,110,116}).compact_done (tigerbeetle)
            tree.manifest.compact(compact_manifest_callback);
                                 ^
/home/phil/tb/tigerbeetle/src/lsm/tree.zig:732:67: 0x4293aa in lsm.tree.TreeType(lsm.table.TableType(lsm.composite_key.CompositeKey(u64),lsm.composite_key.Value,lsm.composite_key.CompositeKey(u64).compare_keys,lsm.composite_key.CompositeKey(u64).key_from_value,(struct lsm.composite_key.CompositeKey(u64) constant),lsm.composite_key.CompositeKey(u64).tombstone,lsm.composite_key.CompositeKey(u64).tombstone_from_key),storage.Storage,[]const u8{84,114,97,110,115,102,101,114,46,97,109,111,117,110,116}).compact_tick_done (tigerbeetle)
            if (tree.compaction_io_pending == 0) tree.compact_done();
                                                                  ^
/home/phil/tb/tigerbeetle/src/lsm/tree.zig:702:35: 0x447701 in lsm.tree.TreeType(lsm.table.TableType(lsm.composite_key.CompositeKey(u64),lsm.composite_key.Value,lsm.composite_key.CompositeKey(u64).compare_keys,lsm.composite_key.CompositeKey(u64).key_from_value,(struct lsm.composite_key.CompositeKey(u64) constant),lsm.composite_key.CompositeKey(u64).tombstone,lsm.composite_key.CompositeKey(u64).tombstone_from_key),storage.Storage,[]const u8{84,114,97,110,115,102,101,114,46,97,109,111,117,110,116}).compact_tick_callback_table_immutable (tigerbeetle)
            tree.compact_tick_done();
                                  ^
/home/phil/tb/tigerbeetle/src/lsm/compaction.zig:389:21: 0x47d732 in lsm.compaction.CompactionType(lsm.table.TableType(lsm.composite_key.CompositeKey(u64),lsm.composite_key.Value,lsm.composite_key.CompositeKey(u64).compare_keys,lsm.composite_key.CompositeKey(u64).key_from_value,(struct lsm.composite_key.CompositeKey(u64) constant),lsm.composite_key.CompositeKey(u64).tombstone,lsm.composite_key.CompositeKey(u64).tombstone_from_key),storage.Storage,lsm.table_immutable.TableImmutableIteratorType).cpu_merge_start (tigerbeetle)
            callback(compaction);
                    ^
/home/phil/tb/tigerbeetle/src/lsm/compaction.zig:358:71: 0x45daa5 in lsm.compaction.CompactionType(lsm.table.TableType(lsm.composite_key.CompositeKey(u64),lsm.composite_key.Value,lsm.composite_key.CompositeKey(u64).compare_keys,lsm.composite_key.CompositeKey(u64).key_from_value,(struct lsm.composite_key.CompositeKey(u64) constant),lsm.composite_key.CompositeKey(u64).tombstone,lsm.composite_key.CompositeKey(u64).tombstone_from_key),storage.Storage,lsm.table_immutable.TableImmutableIteratorType).io_finish (tigerbeetle)
            if (compaction.io_pending == 0) compaction.cpu_merge_start();
                                                                      ^
/home/phil/tb/tigerbeetle/src/lsm/compaction.zig:325:42: 0x47d540 in lsm.compaction.write_callback.callback (tigerbeetle)
                    _compaction.io_finish();
                                         ^
/home/phil/tb/tigerbeetle/src/lsm/grid.zig:330:37: 0x4e7a2c in lsm.grid.GridType(storage.Storage).write_block_callback (tigerbeetle)
            completed_write.callback(completed_write);
                                    ^
/home/phil/tb/tigerbeetle/src/storage.zig:310:27: 0x32c90c in storage.Storage.on_write (tigerbeetle)
            write.callback(write);
                          ^
/home/phil/tb/tigerbeetle/src/io/linux.zig:860:29: 0x32cb40 in io.linux.struct:858:25.wrapper (tigerbeetle)
                    callback(
                            ^
/home/phil/tb/tigerbeetle/src/io/linux.zig:460:40: 0x30ff58 in io.linux.Completion.complete (tigerbeetle)
                    completion.callback(completion.context, completion, &result);
                                       ^
/home/phil/tb/tigerbeetle/src/io/linux.zig:110:64: 0x30dae8 in io.linux.IO.flush (tigerbeetle)
            while (copy.pop()) |completion| completion.complete();
                                                               ^
/home/phil/tb/tigerbeetle/src/io/linux.zig:90:27: 0x2b2972 in io.linux.IO.run_for_ns (tigerbeetle)
            try self.flush(1, &timeouts, &etime);
                          ^
/home/phil/tb/tigerbeetle/src/main.zig:152:38: 0x29bd44 in Command.start (tigerbeetle)
            try command.io.run_for_ns(config.tick_ms * std.time.ns_per_ms);
                                     ^
/home/phil/tb/tigerbeetle/src/main.zig:45:44: 0x293708 in main (tigerbeetle)
        .start => |*args| try Command.start(allocator, args.addresses, args.memory, args.path),
                                           ^
/home/phil/tb/tigerbeetle/zig/lib/std/start.zig:561:37: 0x28cb4a in std.start.callMain (tigerbeetle)
            const result = root.main() catch |err| {
                                    ^
/home/phil/tb/tigerbeetle/zig/lib/std/start.zig:495:12: 0x26d89e in std.start.callMainWithArgs (tigerbeetle)
    return @call(.{ .modifier = .always_inline }, callMain, .{});
           ^
/home/phil/tb/tigerbeetle/zig/lib/std/start.zig:409:17: 0x26c936 in std.start.posixCallMainAndExit (tigerbeetle)
    std.os.exit(@call(.{ .modifier = .always_inline }, callMainWithArgs, .{ argc, argv, envp }));
                ^
/home/phil/tb/tigerbeetle/zig/lib/std/start.zig:322:5: 0x26c742 in std.start._start (tigerbeetle)
    @call(.{ .modifier = .never_inline }, posixCallMainAndExit, .{});
    ^
[1]    2523663 IOT instruction (core dumped)  ./tigerbeetle start --addresses=3000 0_0.tigerbeetle
```

**Comment by jorangreef** [comment]
@sentientwaffle this should be fixed by your https://github.com/tigerbeetledb/tigerbeetle/pull/180

**Comment by sentientwaffle** [comment]
Yes, good catch. @eatonphil after the node client is [updated](https://github.com/tigerbeetledb/tigerbeetle-node/pull/36) can you confirm that this is fixed?

**Comment by batiati** [comment]
I tested it using the Java client:

**1.** The same use case as [the original one](https://gist.github.com/eatonphil/3bceb01e603ea0edcae16217c592c2fb): submitting 100 million batches, each one with 1 transaction.

The server doesn't panic on this assert anymore, but instead, I got a panic on the client side:
```
debug(client): 161962776220313661351056320762770941043: on_reply: user_data=2589905546149477160276484050968128 request=70 size=128 create_transfers
debug(vsr): 161962776220313661351056320762770941043: request_timeout stopped
thread 492081 panic: attempt to use null value
/home/batiati/tigerbeetle-java/src/zig/lib/tigerbeetle/src/message_pool.zig:132:39: 0x7fb1098e21f9 in message_pool.MessagePool.get_message (tb_client)
        const message = pool.free_list.?;
                                      ^
/home/batiati/tigerbeetle-java/src/zig/lib/tigerbeetle/src/message_bus.zig:340:40: 0x7fb1098d6847 in message_bus.MessageBusType(vsr.ProcessType.client).get_message (tb_client)
            return bus.pool.get_message();
                                       ^
/home/batiati/tigerbeetle-java/src/zig/lib/tigerbeetle/src/message_bus.zig:825:52: 0x7fb1098f92b9 in message_bus.Connection.get_recv_message_and_recv (tb_client)
                const new_message = bus.get_message();
                                                   ^
/home/batiati/tigerbeetle-java/src/zig/lib/tigerbeetle/src/message_bus.zig:666:57: 0x7fb109900323 in message_bus.Connection.parse_message (tb_client)
                    connection.get_recv_message_and_recv(bus);
                                                        ^
/home/batiati/tigerbeetle-java/src/zig/lib/tigerbeetle/src/message_bus.zig:656:48: 0x7fb1098ff159 in message_bus.Connection.parse_messages (tb_client)
                while (connection.parse_message(bus)) |message| {
                                               ^
/home/batiati/tigerbeetle-java/src/zig/lib/tigerbeetle/src/message_bus.zig:888:42: 0x7fb1098ff021 in message_bus.Connection.on_recv (tb_client)
                connection.parse_messages(bus);
                                         ^
/home/batiati/tigerbeetle-java/src/zig/lib/tigerbeetle/src/io/linux.zig:722:29: 0x7fb1098ff244 in io.linux.struct:720:25.wrapper (tb_client)
                    callback(
                            ^
/home/batiati/tigerbeetle-java/src/zig/lib/tigerbeetle/src/io/linux.zig:381:40: 0x7fb1098c7048 in io.linux.Completion.complete (tb_client)
                    completion.callback(completion.context, completion, &result);
                                       ^
/home/batiati/tigerbeetle-java/src/zig/lib/tigerbeetle/src/io/linux.zig:110:64: 0x7fb1098c5551 in io.linux.IO.flush (tb_client)
            while (copy.pop()) |completion| completion.complete();
                                                               ^
/home/batiati/tigerbeetle-java/src/zig/lib/tigerbeetle/src/io/linux.zig:90:27: 0x7fb1098c8830 in io.linux.IO.run_for_ns (tb_client)
            try self.flush(1, &timeouts, &etime);
                          ^
/home/batiati/tigerbeetle-java/src/zig/lib/tigerbeetle/src/c/tb_client/thread.zig:211:35: 0x7fb1098c839e in c.tb_client.thread.ThreadType(state_machine.StateMachineType(storage.Storage),message_bus.MessageBusType(vsr.ProcessType.client)).run (tb_client)
                self.io.run_for_ns(config.tick_ms * std.time.ns_per_ms) catch |err| {
                                  ^
/home/batiati/tigerbeetle-java/src/zig/lib/tigerbeetle/zig/lib/std/Thread.zig:359:13: 0x7fb1098d889f in std.Thread.callFn (tb_client)
            @call(.{}, f, args);
            ^
/home/batiati/tigerbeetle-java/src/zig/lib/tigerbeetle/zig/lib/std/Thread.zig:633:30: 0x7fb1098d8352 in std.Thread.Instance.entryFn (tb_client)
                return callFn(f, args_ptr.*);
                             ^
???:?:?: 0x7fb13cd79e2c in ??? (???)
[1]    492055 IOT instruction (core dumped)  /usr/bin/env /usr/lib/jvm/java-11-openjdk-11.0.16.0.8-1.fc36.x86_64/bin/java
```
**2.** Not exactly the same case, but related: submitting 100 batches, each with a full-length message body (8191 transfers).

The server stopped for a couple of seconds, and then I got a server panic in another location.
Before #180 it panicked at the exact location Phil reported.
```
./tigerbeetle start --addresses=3001 file.log
info(io): opening "file.log"...
info(main): 0: cluster=0: listening on 127.0.0.1:3001
info(message_bus): connection from client 107378105930894858831681503437208956857
thread 495475 panic: reached unreachable code
/home/batiati/tigerbeetle-java/src/zig/lib/tigerbeetle/zig/lib/std/debug.zig:225:14: 0x26ad4b in std.debug.assert (tigerbeetle)
    if (!ok) unreachable; // assertion failure
             ^
/home/batiati/tigerbeetle-java/src/zig/lib/tigerbeetle/src/lsm/table_iterator.zig:270:23: 0x2b818f in lsm.table_iterator.TableIteratorType(lsm.table.TableType(lsm.composite_key.CompositeKey(u64),lsm.composite_key.Value,lsm.composite_key.CompositeKey(u64).compare_keys,lsm.composite_key.CompositeKey(u64).key_from_value,(struct lsm.composite_key.CompositeKey(u64) constant),lsm.composite_key.CompositeKey(u64).tombstone,lsm.composite_key.CompositeKey(u64).tombstone_from_key),storage.Storage).peek (tigerbeetle)
                assert(it.data_block_index == Table.index_data_blocks_used(it.index_block));
                      ^
/home/batiati/tigerbeetle-java/src/zig/lib/tigerbeetle/src/lsm/level_iterator.zig:296:36: 0x2b7deb in lsm.level_iterator.LevelIteratorType(lsm.table.TableType(lsm.composite_key.CompositeKey(u64),lsm.composite_key.Value,lsm.composite_key.CompositeKey(u64).compare_keys,lsm.composite_key.CompositeKey(u64).key_from_value,(struct lsm.composite_key.CompositeKey(u64) constant),lsm.composite_key.CompositeKey(u64).tombstone,lsm.composite_key.CompositeKey(u64).tombstone_from_key),storage.Storage).pop (tigerbeetle)
            if (table_iterator.peek() == null) it.tables.advance_head();
                                   ^
/home/batiati/tigerbeetle-java/src/zig/lib/tigerbeetle/src/lsm/compaction.zig:162:51: 0x2b7a95 in lsm.compaction.MergeStreamSelector.pop (tigerbeetle)
                    1 => compaction.iterator_b.pop(),
                                                  ^
/home/batiati/tigerbeetle-java/src/zig/lib/tigerbeetle/src/lsm/k_way_merge.zig:103:37: 0x551721 in lsm.k_way_merge.KWayMergeIterator(lsm.compaction.CompactionType(lsm.table.TableType(lsm.composite_key.CompositeKey(u64),lsm.composite_key.Value,lsm.composite_key.CompositeKey(u64).compare_keys,lsm.composite_key.CompositeKey(u64).key_from_value,(struct lsm.composite_key.CompositeKey(u64) constant),lsm.composite_key.CompositeKey(u64).tombstone,lsm.composite_key.CompositeKey(u64).tombstone_from_key),storage.Storage,lsm.table_immutable.TableImmutableIteratorType),lsm.composite_key.CompositeKey(u64),lsm.composite_key.Value,lsm.composite_key.CompositeKey(u64).key_from_value,lsm.composite_key.CompositeKey(u64).compare_keys,2,lsm.compaction.MergeStreamSelector.peek,lsm.compaction.MergeStreamSelector.pop,lsm.compaction.MergeStreamSelector.precedence).pop_internal (tigerbeetle)
            const value = stream_pop(it.context, root);
                                    ^
/home/batiati/tigerbeetle-java/src/zig/lib/tigerbeetle/src/lsm/k_way_merge.zig:80:35: 0x517683 in lsm.k_way_merge.KWayMergeIterator(lsm.compaction.CompactionType(lsm.table.TableType(lsm.composite_key.CompositeKey(u64),lsm.composite_key.Value,lsm.composite_key.CompositeKey(u64).compare_keys,lsm.composite_key.CompositeKey(u64).key_from_value,(struct lsm.composite_key.CompositeKey(u64) constant),lsm.composite_key.CompositeKey(u64).tombstone,lsm.composite_key.CompositeKey(u64).tombstone_from_key),storage.Storage,lsm.table_immutable.TableImmutableIteratorType),lsm.composite_key.CompositeKey(u64),lsm.composite_key.Value,lsm.composite_key.CompositeKey(u64).key_from_value,lsm.composite_key.CompositeKey(u64).compare_keys,2,lsm.compaction.MergeStreamSelector.peek,lsm.compaction.MergeStreamSelector.pop,lsm.compaction.MergeStreamSelector.precedence).pop (tigerbeetle)
            while (it.pop_internal()) |value| {
                                  ^
/home/batiati/tigerbeetle-java/src/zig/lib/tigerbeetle/src/lsm/compaction.zig:485:49: 0x4c9503 in lsm.compaction.CompactionType(lsm.table.TableType(lsm.composite_key.CompositeKey(u64),lsm.composite_key.Value,lsm.composite_key.CompositeKey(u64).compare_keys,lsm.composite_key.CompositeKey(u64).key_from_value,(struct lsm.composite_key.CompositeKey(u64) constant),lsm.composite_key.CompositeKey(u64).tombstone,lsm.composite_key.CompositeKey(u64).tombstone_from_key),storage.Storage,lsm.table_immutable.TableImmutableIteratorType).cpu_merge (tigerbeetle)
                const value = merge_iterator.pop() orelse break;
                                                ^
/home/batiati/tigerbeetle-java/src/zig/lib/tigerbeetle/src/lsm/compaction.zig:455:37: 0x486b6e in lsm.compaction.CompactionType(lsm.table.TableType(lsm.composite_key.CompositeKey(u64),lsm.composite_key.Value,lsm.composite_key.CompositeKey(u64).compare_keys,lsm.composite_key.CompositeKey(u64).key_from_value,(struct lsm.composite_key.CompositeKey(u64) constant),lsm.composite_key.CompositeKey(u64).tombstone,lsm.composite_key.CompositeKey(u64).tombstone_from_key),storage.Storage,lsm.table_immutable.TableImmutableIteratorType).cpu_merge_start (tigerbeetle)
                compaction.cpu_merge();
                                    ^
/home/batiati/tigerbeetle-java/src/zig/lib/tigerbeetle/src/lsm/compaction.zig:434:71: 0x467195 in lsm.compaction.CompactionType(lsm.table.TableType(lsm.composite_key.CompositeKey(u64),lsm.composite_key.Value,lsm.composite_key.CompositeKey(u64).compare_keys,lsm.composite_key.CompositeKey(u64).key_from_value,(struct lsm.composite_key.CompositeKey(u64) constant),lsm.composite_key.CompositeKey(u64).tombstone,lsm.composite_key.CompositeKey(u64).tombstone_from_key),storage.Storage,lsm.table_immutable.TableImmutableIteratorType).io_finish (tigerbeetle)
            if (compaction.io_pending == 0) compaction.cpu_merge_start();
                                                                      ^
/home/batiati/tigerbeetle-java/src/zig/lib/tigerbeetle/src/lsm/compaction.zig:338:33: 0x466340 in lsm.compaction.CompactionType(lsm.table.TableType(lsm.composite_key.CompositeKey(u64),lsm.composite_key.Value,lsm.composite_key.CompositeKey(u64).compare_keys,lsm.composite_key.CompositeKey(u64).key_from_value,(struct lsm.composite_key.CompositeKey(u64) constant),lsm.composite_key.CompositeKey(u64).tombstone,lsm.composite_key.CompositeKey(u64).tombstone_from_key),storage.Storage,lsm.table_immutable.TableImmutableIteratorType).iterator_b_io_callback (tigerbeetle)
            compaction.io_finish();
                                ^
/home/batiati/tigerbeetle-java/src/zig/lib/tigerbeetle/src/lsm/level_iterator.zig:247:28: 0x486840 in lsm.level_iterator.LevelIteratorType(lsm.table.TableType(lsm.composite_key.CompositeKey(u64),lsm.composite_key.Value,lsm.composite_key.CompositeKey(u64).compare_keys,lsm.composite_key.CompositeKey(u64).key_from_value,(struct lsm.composite_key.CompositeKey(u64) constant),lsm.composite_key.CompositeKey(u64).tombstone,lsm.composite_key.CompositeKey(u64).tombstone_from_key),storage.Storage).table_iterator_callback (tigerbeetle)
                it.callback(it);
                           ^
/home/batiati/tigerbeetle-java/src/zig/lib/tigerbeetle/src/lsm/table_iterator.zig:229:29: 0x4ca139 in lsm.table_iterator.TableIteratorType(lsm.table.TableType(lsm.composite_key.CompositeKey(u64),lsm.composite_key.Value,lsm.composite_key.CompositeKey(u64).compare_keys,lsm.composite_key.CompositeKey(u64).key_from_value,(struct lsm.composite_key.CompositeKey(u64) constant),lsm.composite_key.CompositeKey(u64).tombstone,lsm.composite_key.CompositeKey(u64).tombstone_from_key),storage.Storage).on_read (tigerbeetle)
                it.read_done(it);
                            ^
/home/batiati/tigerbeetle-java/src/zig/lib/tigerbeetle/src/lsm/grid.zig:414:30: 0x30c53e in lsm.grid.GridType(storage.Storage).start_read (tigerbeetle)
                read.callback(read, block);
                             ^
/home/batiati/tigerbeetle-java/src/zig/lib/tigerbeetle/src/lsm/grid.zig:176:32: 0x2bebcb in lsm.grid.GridType(storage.Storage).tick (tigerbeetle)
                grid.start_read(read);
                               ^
/home/batiati/tigerbeetle-java/src/zig/lib/tigerbeetle/src/vsr/replica.zig:552:27: 0x2b3d81 in vsr.replica.ReplicaType(state_machine.StateMachineType(storage.Storage),message_bus.MessageBusType(vsr.ProcessType.replica),storage.Storage,time.Time).tick (tigerbeetle)
            self.grid.tick();
                          ^
/home/batiati/tigerbeetle-java/src/zig/lib/tigerbeetle/src/main.zig:151:25: 0x29d899 in Command.start (tigerbeetle)
            replica.tick();
                        ^
/home/batiati/tigerbeetle-java/src/zig/lib/tigerbeetle/src/main.zig:45:44: 0x29526b in main (tigerbeetle)
        .start => |*args| try Command.start(allocator, args.addresses, args.memory, args.path),
                                           ^
/home/batiati/tigerbeetle-java/src/zig/lib/tigerbeetle/zig/lib/std/start.zig:561:37: 0x28e6ea in std.start.callMain (tigerbeetle)
            const result = root.main() catch |err| {
                                    ^
/home/batiati/tigerbeetle-java/src/zig/lib/tigerbeetle/zig/lib/std/start.zig:495:12: 0x26f4ae in std.start.callMainWithArgs (tigerbeetle)
    return @call(.{ .modifier = .always_inline }, callMain, .{});
           ^
/home/batiati/tigerbeetle-java/src/zig/lib/tigerbeetle/zig/lib/std/start.zig:409:17: 0x26e546 in std.start.posixCallMainAndExit (tigerbeetle)
    std.os.exit(@call(.{ .modifier = .always_inline }, callMainWithArgs, .{ argc, argv, envp }));
                ^
/home/batiati/tigerbeetle-java/src/zig/lib/tigerbeetle/zig/lib/std/start.zig:322:5: 0x26e352 in std.start._start (tigerbeetle)
    @call(.{ .modifier = .never_inline }, posixCallMainAndExit, .{});
    ^
[1]    495475 IOT instruction (core dumped)  ./tigerbeetle start --addresses=3001 file.log

```

**Comment by sentientwaffle** [comment]
Thanks @batiati, this is great!

1. Possible fix: https://github.com/tigerbeetledb/tigerbeetle/pull/182
2. Also found via fuzzer! https://github.com/tigerbeetledb/tigerbeetle/issues/188

**Comment by batiati** [comment]
Hi @sentientwaffle 
It looks like I'm hitting #187 after your fix #182.

Well done fuzzer!!! 🚀

**Comment by sentientwaffle** [comment]
Closing in favor of https://github.com/tigerbeetledb/tigerbeetle/issues/188; the other two bugs mentioned in this thread are fixed.
